### ISSUE TITLE
Alacritty does not render on startup until layout is changed

### ISSUE THREAD

**Issue by LogicalOverflow** [pinnacle-comp/pinnacle#202]
When I start alacritty in Pinnacle, it does not show up (most of the time, like 9/10) until I change the layout. Sometimes a single cycling of the layout is not enough (i.e. I need to cycle a few times until alacritty shows up). By "not show up", I mean it gets space in the layout, but this space remains empty (i.e. I see the wallpaper).  This also applies to the alacritty instance spawned at the end of the default config.

If the invisible alacritty window is the only window, I am even able to type in the terminal. Once I cycle the layout, alacritty (with the entered text) shows up.

This seems to be an alacritty-specific issue; at least I wasn't able to reproduce it with other things.

Might be related to #104 and/or #173, not sure.

I'll try to look into this a bit more in the coming days to figure out what's going on.

I am using commit 1ba636f, the default rust config, rust 1.77.2, NixOS (23.11), and the winit backend (on sway).

**Comment by Ottatop** [comment]
Hm interesting, I can't reproduce this.

The reason you can type in Alacritty if it's the only window is because it has keyboard focus. This means you should be able to toggle floating/maximized/fullscreen. Can you try that on the invisible Alacritty window and see if that also makes it appear?

Also, to clarify, is this *just* the first Alacritty or do subsequent spawned Alacrittys also have the same problem?

Also also, can you attach any relevant/recent logs in `~/.local/state/pinnacle`? Maybe there's something useful in them.

**Comment by LogicalOverflow** [comment]
> Can you try that on the invisible Alacritty window and see if that also makes it appear?

Toggling fullscreen does not make Alacritty visible, but is able to full screen the hidden window: If I open a single Alacritty (which is then invisible), then toggle full-screen and then cycle the layout, a full-screened Alacritty appears. So pretty much the same behavior as typing into the invisible window.

> Also, to clarify, is this just the first Alacritty or do subsequent spawned Alacrittys also have the same problem?

This effects (almost) all Alacritty instances; in rare cases, everything "just works" randomly. I also tried starting Alacritty from a terminal inside Pinnacle, instead of over the API, and the issue also occurs there.

> Also also, can you attach any relevant/recent logs in ~/.local/state/pinnacle? Maybe there's something useful in them.

The original logs are huge, as I turned on trace-level logging trying to figure out whats going on. As the bug is quite consistent on my system, here's a newly created log:  [pinnacle-alacritty-invisible.log](https://github.com/pinnacle-comp/pinnacle/files/14965698/pinnacle-alacritty-invisible.log)
In it I do the following:
1. Start Pinnacle
2. Cycle the layout once to show the Alacritty started at the end of the default config
3. Open a second Alacritty
4. Cycle the layout until this second Alacritty shows up. This takes 4 layout cycles.
5. Close all the terminals (using the `exit` command, i.e. not through Pinnacle) and then Pinnacle

For these logs, I added two `debug!` events, `LayoutService.layout.Geometries` and `LayoutService.layout.ExplicitLayout`, emitted whenever a streamed layout request is received to make them show up clearly in the log. (This is the only change compared to commit 1ba636f)

**Comment by LogicalOverflow** [comment]
Okay, I did some digging. The issue seems to be in `State::update_windows_with_geometries`, specifically with https://github.com/pinnacle-comp/pinnacle/blob/1ba636f39f5203a15aba39504d3fd6c1c71fdf6e/src/layout.rs#L113-L120 and https://github.com/pinnacle-comp/pinnacle/blob/1ba636f39f5203a15aba39504d3fd6c1c71fdf6e/src/layout.rs#L132-L134

Alacritty usually has pending changes (i.e. `pending` is true), when `update_windows_with_geometries` is called. As such, it lands in the `peinding_wins` vector and the changes are committed. Problem is, only the windows in `non_pending_wins` are mapped into the space with their locations. In fact, the vector `pending_wins` is only populated but never read from.

I found to ways to resolve the issue:
- Moving the code in the `else` case out of it (i.e. adding all windows to `non_pending_wins`), which calls `self.space.map_element` also for the windows with pending changes.
- Implementing `ack_configure` in the `XdgShellHandler` implementation for `State` and calling `self.space.map_element` for the acknowledging window here.

I am not sure which is the better solution here; I don't know enough about the wayland protocols etc. My (uninformed) intuition leans towards the `ack_configure` version.

**Comment by LogicalOverflow** [comment]
Okay, I did some more digging and it seems a bit more complicated: Currently, the code does the mapping for windows with pending position changes when the client commits (at least for non-sync windows)
 https://github.com/pinnacle-comp/pinnacle/blob/1ba636f39f5203a15aba39504d3fd6c1c71fdf6e/src/handlers.rs#L128-L135

The issue occurs, because Alacratty does not commit, it only calls `ack_configure` for the pending changes. This seems to be related to the initial offscreen-mapping to `(1000000, 0)`. My current hypothesis is the following:
1. Alacritty gets mapped out of bounds  and therefore renders nothing
2. It receives a configure, which it acknowledges, but does not commit as it thinks there is no change, as the mapping is still off-screen and there is nothing to render anyways
3. As we never get a commit, Pinnacle never gets to re-map the surface on-screen, leading to a kind of deadlock leaving Alacritty off-screen

With this, I found two more ways to resolve this:
- Moving the initial mapping to `(0, 0)` also solve this issue, but leads to a quick flicker of Alacritty with its default dimensions.
- Sending the frame event to Alacritty triggers a commit, resolving the deadlock.

I am not sure, what the correct point for the remapping actually is. Should it wait for a configure and commit or not? 
- If it should wait, I think `update_windows_with_geometries` also should not re-map any windows and only the commit handler should do the re-mappings. Not sure how to inform clients that Pinnacle needs a new frame tho.
- If it should not wait, I think it makes sense to do all re-mappings in `update_windows_with_geometries`. This feels like the easier way to do this. Not sure if that can lead to some flickering, for windows being moved and re-sized at the same time, with them being moved "instantly", but only re-sized once the client acknowledges the re-size.

**Comment by Ottatop** [comment]
> If it should wait, I think `update_windows_with_geometries` also should not re-map any windows and only the commit handler should do the re-mappings.

The problem with this is that some layout applications don't change the size of a window, meaning it won't get a commit. This is why non-pending windows are partitioned off and mapped in `update_windows_with_geometries`.

> If it should not wait, I think it makes sense to do all re-mappings in `update_windows_with_geometries`. This feels like the easier way to do this. Not sure if that can lead to some flickering, for windows being moved and re-sized at the same time, with them being moved "instantly", but only re-sized once the client acknowledges the re-size.

I'm *fairly* certain that there's a possibility of time being between a configure ACK and the actual commit, meaning that there would be the flickering you mentioned.

Ideally there would be some sort of transactional layout system that waits for everything to ACK then triggers a commit/maps if there are no pending changes. Also ideal would be for me to request layout geometries but not actually apply the layout until later, allowing me to set the initial size and bounds for new windows in `XdgShellHandler::new_toplevel`. This would most likely stop the flickering you mention when moving the initial map to (0, 0).

As a stopgap, can you apply the following diff and see if it's also a workaround to the problem?
```diff
diff --git a/src/layout.rs b/src/layout.rs
index 60167a38d..f39c35446 100644
--- a/src/layout.rs
+++ b/src/layout.rs
@@ -1,6 +1,9 @@
 // SPDX-License-Identifier: GPL-3.0-or-later
 
-use std::collections::{HashMap, HashSet};
+use std::{
+    collections::{HashMap, HashSet},
+    time::Duration,
+};
 
 use pinnacle_api_defs::pinnacle::layout::v0alpha1::{layout_request::Geometries, LayoutResponse};
 use smithay::{
@@ -133,6 +136,12 @@ impl State {
             self.space.map_element(window, loc, false);
         }
 
+        for (win, _serial) in pending_wins {
+            win.send_frame(output, self.clock.now(), Some(Duration::ZERO), |_, _| {
+                Some(output.clone())
+            });
+        }
+
         self.fixup_z_layering();
     }
```
This is slightly better than mapping directly in `update_windows_with_geometries` if it works. Granted it reeks of me being too lazy to actually fix the underlying problem of mapping to (1000000, 0), but hey, stopgap.

> In fact, the vector `pending_wins` is only populated but never read from.

Classic me being a bad programmer :trollface: 

Also thanks for taking the time to look into this!

**Comment by LogicalOverflow** [comment]
> As a stopgap, can you apply the following diff and see if it's also a workaround to the problem?

Yes, that fix works on my setup.

> I'm fairly certain that there's a possibility of time being between a configure ACK and the actual commit, meaning that there would be the flickering you mentioned.

From what I have read throughout the day, that matches my understanding of "The ack only really matters, once the commit comes in". (Clients are even allowed to do multiple acks before a commit).

Also, positions of windows seem, for the most part, to be a compositor-only thing. As such, moving windows without coordinating with the client feels like the normal behavior. But things get complicated once you try to do moving and re-sizing at the same time.

(I am not experienced with the wayland protocol btw, I have just played around with it a few times etc., so I don't know how correct this really is.)

(Another related thing I stumbled across: Once a window has ack'ed the configuration, it no longer has pending changes, even if no commit happens. As such, if a layout is applied in the time window between the ack and commit (which does not itself change the configuration), it will trigger the re-mapping (i.e. be similar to doing the re-mapping when Pinnacle receives the ack). This is also why cycling the layout fixes Alacritty, I think.)

Overall, I think the stopgap solution is fine to solve this particular issue for now. Clean solutions for the underlying problems ("not having the layout geometry when creating windows" and "windows don't get moved if they don't commit a frame and have pending changes") are probably a bigger things for a later date; maybe they can get their own issues at some point.

> Also thanks for taking the time to look into this!

You're welcome! You are pretty much writing the compositor I was thinking of building myself (or at least a very similar one). Now I am trying to help out here instead of spending time trying to building my own compositor :D (and am quite excited about the whole thing, tbh)
